File: src/main/java/com/elixir/mars/repository/IsaJsonGraphLookup.java
package com.elixir.mars.repository;

import com.elixir.mars.repository.models.isa.DataFile;
import com.elixir.mars.repository.models.isa.Input;
import com.elixir.mars.repository.models.isa.Materials;
import com.elixir.mars.repository.models.isa.OtherMaterial;
import com.elixir.mars.repository.models.isa.Output;
import com.elixir.mars.repository.models.isa.ProcessSequence;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Shared ISA-JSON graph helpers for repository service conversions.
 *
 * <p>Repository metadata and receipt conversion all need to resolve the same process/data-file
 * lineage. Keeping the lookup rules here avoids subtle drift, especially around ISA's
 * #data_file/... versus #data/... ID forms.
 */
public final class IsaJsonGraphLookup {
  private static final String DATA_FILE_ID_PREFIX = "#data_file/";
  private static final String DATA_ID_PREFIX = "#data/";

  private IsaJsonGraphLookup() {}

  /**
   * Finds the process that produced the given output material or data file.
   *
   * <p>ISA process outputs and assay data files can use different ID prefixes for the same
   * object, so data file IDs are normalized before comparison.
   */
  public static ProcessSequence findProcessByOutputId(
      final List<ProcessSequence> processSequence, final String outputId) {
    if (processSequence == null || outputId == null) {
      return null;
    }

    final String normalizedOutputId = normalizeDataFileId(outputId);

    for (final ProcessSequence process : processSequence) {
      if (process == null || process.getOutputs() == null) {
        continue;
      }

      for (final Output output : process.getOutputs()) {
        if (output == null || output.getId() == null) {
          continue;
        }

        if (normalizeDataFileId(output.getId()).equals(normalizedOutputId)) {
          return process;
        }
      }
    }

    return null;
  }

  /**
   * Finds the first assay other material referenced by a process input.
   *
   * <p>In ISA, a process input stores only an item ID. This helper resolves that ID against the
   * assay materials collection and returns the matching other material, if one exists.
   */
  public static OtherMaterial findOtherMaterialFromProcessInput(
      final ProcessSequence process, final Materials materials) {
    if (process == null || process.getInputs() == null) {
      return null;
    }

    final Map<String, OtherMaterial> otherMaterialsById = buildOtherMaterialsById(materials);
    for (final Input input : process.getInputs()) {
      if (input == null || input.getId() == null) {
        continue;
      }

      final OtherMaterial otherMaterial = otherMaterialsById.get(input.getId());
      if (otherMaterial != null) {
        return otherMaterial;
      }
    }

    return null;
  }

  /**
   * Finds assay data files referenced by a process's outputs.
   *
   * <p>Process outputs can point to data files using the process-level ID form. This helper
   * normalizes those IDs before matching them against the assay data file list.
   */
  public static List<DataFile> findDataFilesFromProcessOutputs(
      final ProcessSequence process, final List<DataFile> assayDataFiles) {
    final List<DataFile> dataFiles = new ArrayList<>();
    if (process == null || process.getOutputs() == null || assayDataFiles == null) {
      return dataFiles;
    }

    for (final Output output : process.getOutputs()) {
      if (output == null || output.getId() == null) {
        continue;
      }

      final String normalizedOutputId = normalizeDataFileId(output.getId());
      for (final DataFile dataFile : assayDataFiles) {
        if (dataFile == null || dataFile.getId() == null) {
          continue;
        }

        if (normalizeDataFileId(dataFile.getId()).equals(normalizedOutputId)) {
          dataFiles.add(dataFile);
        }
      }
    }

    return dataFiles;
  }

  /** Builds a lookup map of assay other materials keyed by their ISA IDs. */
  public static Map<String, OtherMaterial> buildOtherMaterialsById(final Materials materials) {
    final Map<String, OtherMaterial> otherMaterialsById = new HashMap<>();
    if (materials == null || materials.getOtherMaterials() == null) {
      return otherMaterialsById;
    }

    for (final OtherMaterial otherMaterial : materials.getOtherMaterials()) {
      if (otherMaterial != null && otherMaterial.getId() != null) {
        otherMaterialsById.put(otherMaterial.getId(), otherMaterial);
      }
    }

    return otherMaterialsById;
  }

  /** Treat ISA #data_file/... and process #data/... references as equivalent. */
  public static String normalizeDataFileId(final String id) {
    return id == null ? null : id.replace(DATA_FILE_ID_PREFIX, DATA_ID_PREFIX);
  }
}

File: python/mars_repository/isa_json_graph_lookup.py
from typing import Dict, List, Optional

from .models.isa.data_file import DataFile
from .models.isa.materials import Materials
from .models.isa.other_material import OtherMaterial
from .models.isa.output import Output
from .models.isa.process_sequence import ProcessSequence

DATA_FILE_ID_PREFIX = "#data_file/"
DATA_ID_PREFIX = "#data/"


def normalize_data_file_id(id: Optional[str]) -> Optional[str]:
    if id is None:
        return None
    return id.replace(DATA_FILE_ID_PREFIX, DATA_ID_PREFIX)


def build_other_materials_by_id(
    materials: Optional[Materials],
) -> Dict[str, OtherMaterial]:
    other_materials_by_id: Dict[str, OtherMaterial] = {}
    if materials is None or materials.get("other_materials") is None:
        return other_materials_by_id
    for om in materials["other_materials"]:
        if om is not None and om.get("id") is not None:
            other_materials_by_id[om["id"]] = om
    return other_materials_by_id


def find_process_by_output_id(
    process_sequence: Optional[List[ProcessSequence]],
    output_id: Optional[str],
) -> Optional[ProcessSequence]:
    if process_sequence is None or output_id is None:
        return None

    normalized_output_id = normalize_data_file_id(output_id)

    for process in process_sequence:
        if process is None or process.get("outputs") is None:
            continue

        for output in process["outputs"]:
            if output is None or output.get("id") is None:
                continue

            if normalize_data_file_id(output["id"]) == normalized_output_id:
                return process

    return None


def find_other_material_from_process_input(
    process: Optional[ProcessSequence],
    materials: Optional[Materials],
) -> Optional[OtherMaterial]:
    if process is None or process.get("inputs") is None:
        return None

    other_materials_by_id = build_other_materials_by_id(materials)
    for input_item in process["inputs"]:
        if input_item is None or input_item.get("id") is None:
            continue

        other_material = other_materials_by_id.get(input_item["id"])
        if other_material is not None:
            return other_material

    return None


def find_data_files_from_process_outputs(
    process: Optional[ProcessSequence],
    assay_data_files: Optional[List[DataFile]],
) -> List[DataFile]:
    data_files: List[DataFile] = []
    if process is None or process.get("outputs") is None or assay_data_files is None:
        return data_files

    for output in process["outputs"]:
        if output is None or output.get("id") is None:
            continue

        normalized_output_id = normalize_data_file_id(output["id"])
        for data_file in assay_data_files:
            if data_file is None or data_file.get("id") is None:
                continue

            if normalize_data_file_id(data_file["id"]) == normalized_output_id:
                data_files.append(data_file)

    return data_files
